src/profile-logic/transforms.js

```diff
@@ -399,6 +445,143 @@ export function mergeFunction(
+    length: frameTable.length,
+  };
+  const newFuncTable: FuncTable = {
+    address: funcTable.address.slice(),
+    isJS: funcTable.isJS.slice(),
+    name: funcTable.name.slice(),
+    resource: funcTable.resource.slice(),
+    fileName: funcTable.fileName.slice(),
+    lineNumber: funcTable.lineNumber.slice(),
+    length: funcTable.length,
+  };
+  const newStackTable: StackTable = {
+    length: 0,
+    prefix: [],
+    frame: [],
+  };
+  const oldStackToNewStack: Map<
+    IndexIntoStackTable | null,
+    IndexIntoStackTable | null
+  > = new Map();
+  const prefixStackToCollapsedStack: Map<
+    IndexIntoStackTable | null, // prefix stack index
+    IndexIntoStackTable | null // collapsed stack index
+  > = new Map();
+  const stackToCollapsedFrame: Map<
+    IndexIntoStackTable | null,
+    IndexIntoFrameTable | null
+  > = new Map();
+
+  oldStackToNewStack.set(null, null);
+  prefixStackToCollapsedStack.set(null, null);
+  stackToCollapsedFrame.set(null, null);
+
+  for (let stackIndex = 0; stackIndex < stackTable.length; stackIndex++) {
+    const prefix = stackTable.prefix[stackIndex];
+    const frameIndex = stackTable.frame[stackIndex];
+    const funcIndex = frameTable.func[frameIndex];
+    const resourceIndex = funcTable.resource[funcIndex];
+    const newStackPrefix = oldStackToNewStack.get(prefix);
+
+    if (newStackPrefix === undefined) {
+      throw new Error('newStackPrefix must not be undefined');
+    }
+    if (resourceIndex === resourceIndexToCollapse) {
+      // The stack matches this resource.
+      const newPrefixFrame = stackToCollapsedFrame.get(newStackPrefix);
+      if (newPrefixFrame === undefined) {
+        // The prefix is not a collapsed stack. Now check for an existing collapsed
+        // stack at this level. If it exists we can use that.
+        const existingCollapsedStack = prefixStackToCollapsedStack.get(prefix);
+        if (existingCollapsedStack === undefined) {
+          // Create a new collapsed frame.
+
+          // Compute the next indexes
+          const newStackIndex = newStackTable.length++;
+          const newFrameIndex = newFrameTable.length++;
+          const newFuncIndex = newFuncTable.length++;
+          stackToCollapsedFrame.set(newStackIndex, newFrameIndex);
+          oldStackToNewStack.set(stackIndex, newStackIndex);
+          prefixStackToCollapsedStack.set(prefix, newStackIndex);
+
+          // Add the collapsed frame
+          newFrameTable.address.push(frameTable.address[frameIndex]);
+          newFrameTable.category.push(frameTable.category[frameIndex]);
+          newFrameTable.func.push(newFuncIndex);
+          newFrameTable.line.push(frameTable.line[frameIndex]);
+          newFrameTable.implementation.push(
+            frameTable.implementation[frameIndex]
+          );
+          newFrameTable.optimizations.push(
+            frameTable.optimizations[frameIndex]
+          );
+
+          // Add the psuedo-func
```

Comment:
Wouldn't it be enough to have one pseudo func and one pseudo frame per resource? You're adding a new one for every new collapsed stack.

Reply:
Woops, you're correct.